### PR TITLE
Transpose; juxt; identity; compose; pipe; index_on

### PR DESCRIPTION
**Pull request by boxxxie**

### CODE REVIEW

File: lib/kitchensink/function.ex [hunk 1]
```diff
@@ -11,4 +11,35 @@ defmodule KitchenSink.Function do
...
+  def identity(x), do: x
+
+  def compose(functions) when is_list(functions) do
+    functions = Enum.reverse(functions)
+    fn(input) ->
+      Enum.reduce(functions, input, fn (fun, acc) ->
+        fun.(acc)
+      end)
+    end
+  end
```
Comment (kenips) [inline lib/kitchensink/function.ex]: How about:

```
def compose(functions) when is_list(functions) do
  functions
  |> Enum.reverse()
  |> pipe()
end
```

File: lib/kitchensink/function.ex [hunk 2]
```diff
@@ -11,4 +11,35 @@ defmodule KitchenSink.Function do
     fn(arg1, arg2) -> fun.(arg2, arg1) end
   end
 
+  def identity(x), do: x
+
+  def compose(functions) when is_list(functions) do
```
Comment (kenips) [inline lib/kitchensink/function.ex]: Need some @doc

File: lib/kitchensink/function.ex [hunk 3]
```diff
@@ -11,4 +11,35 @@ defmodule KitchenSink.Function do
...
+    functions = Enum.reverse(functions)
+    fn(input) ->
+      Enum.reduce(functions, input, fn (fun, acc) ->
+        fun.(acc)
+      end)
+    end
+  end
+
+  def pipe(functions) when is_list(functions) do
```
Comment (kenips) [inline lib/kitchensink/function.ex]: @doc

File: test/kitchensink_test/function_test.exs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    assert F.identity(1) == 1
+    assert F.identity(%{a: 1}) == %{a: 1}
+  end
+
+  test "compose" do
+    double = fn(x) -> x * 2 end
+    triple = fn(x) -> x * 3 end
+    square = fn(x) -> x * x end
+
+    assert F.compose([square, double, triple]).(3) == 3 * 3 * 2 * 18
```
Comment (kenips) [inline test/kitchensink_test/function_test.exs]: How about `assert F.compose([square, double, triple]).(3) == :math.pow(3 * 3 * 2, 2)`

**Review by kenips — CHANGES_REQUESTED** [review]

File: lib/kitchensink/function.ex (hunk 1 above)
Comment (boxxxie) [reply lib/kitchensink/function.ex]: I made pipe after. I'll change this.

File: lib/kitchensink/list.ex [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  @moduledoc """
+  this module is for List functions
+  """
+
+
+  @doc """
+  takes a list of maps, transforms it into a map of maps with their value being the value_key. basically making a
+  look-up table.
+  """
+  def index_on(list_of_maps, take_keys, value_key) do
```
Comment (kenips) [inline lib/kitchensink/list.ex:12]: not sure about `index_on`. Maybe `to_lookup` is better?

File: test/kitchensink_test/list_test.exs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  doctest KitchenSink.List
+
+  test "index_on" do
+    expected = %{
+      %{a: 1} => 1,
+      %{a: 2} => 2,
+      %{a: 3} => 3
+    }
+
+      input =
```
Comment (kenips) [inline test/kitchensink_test/list_test.exs]: tab!

File: lib/kitchensink/list.ex (hunk 5 above)
Comment (kenips) [reply lib/kitchensink/list.ex:12]: How about adding support for `take_keys` to be one to many.

**Review by kenips — CHANGES_REQUESTED** [review]

File: lib/kitchensink/list.ex (hunk 5 above)
Comment (boxxxie) [reply lib/kitchensink/list.ex:12]: name inspired by underscore

Comment (boxxxie) [reply lib/kitchensink/list.ex:12]: which is indexBy

Comment (boxxxie) [reply lib/kitchensink/list.ex:12]: i think that to_... sounds like it's converting to a proper data structure... this is a generic map. i would like a better name.

Comment (kenips) [reply lib/kitchensink/list.ex:12]: index by then.

**Review by kenips — APPROVED** [review]

**Pull request merged**